graphbolt/src/csc_sampling_graph.cc

```diff
@@ -198,21 +208,60 @@ c10::intrusive_ptr<CSCSamplingGraph> CSCSamplingGraph::LoadFromSharedMemory(
 
 torch::Tensor Pick(
     int64_t offset, int64_t num_neighbors, int64_t fanout, bool replace,
-    const torch::TensorOptions& options) {
+    const torch::TensorOptions& options,
+    const torch::optional<torch::Tensor>& probs) {
```

Comment:
I see we are dropping the original design that passes in a pick UDF. I wonder if that design will make the code cleaner? In fact, we could treat the logic to handle with or without replacement, uniform or non-uniform sampling or parts of the pick UDF logic. For example,

```c++
c10::intrusive_ptr<SampledSubgraph> CSCSamplingGraph::SampleNeighbors(
    const torch::Tensor& nodes, const std::vector<int64_t>& fanouts,
    PickUDF pick_udf) const {
  // The pick UDF will handle the detailed logics to sample on a segment while
  // this function handles the outer parallelism.
  ...

  // the big loop
  torch::parallel_for(0, num_nodes, 32, [&](size_t b, size_t e) {
    for (size_t i = b; i < e; ++i) {
      const auto nid = nodes[i].item<int64_t>();
      TORCH_CHECK(
          nid >= 0 && nid < NumNodes(),
          "The seed nodes' IDs should fall within the range of the graph's "
          "node IDs.");
      const auto offset = indptr_[nid].item<int64_t>();
      const auto num_neighbors = indptr_[nid + 1].item<int64_t>() - offset;

      if (num_neighbors == 0) {
        // Initialization is performed here because all tensors will be
        // concatenated in the master thread, and having an undefined tensor
        // during concatenation can result in a crash.
        picked_neighbors_per_node[i] = torch::tensor({}, indptr_.options());
        continue;
      }

      if (consider_etype) {
        picked_neighbors_per_node[i] = PickEtype(
            offset, num_neighbors, fanouts, indptr_.options(),
            type_per_edge_.value(), pick_udf);  // just pass in the UDF.
      } else {
        picked_neighbors_per_node[i] = Pick(
            offset, num_neighbors, fanouts[0], indptr_.options(),
            pick_udf);  // Just pass-in the UDF
      }
      num_picked_neighbors_per_node[i + 1] =
          picked_neighbors_per_node[i].size(0);
    }
  });  // End of the thread.
  ...
}
```

And then we can just invoke the function with different UDFs for different scenarios (instead of having mingled logics). For example,

```c++
// For uniform sampling
SampleNeighbors(nodes, fanouts, UniformPickUDF);
// For non-uniform sampling
SampleNeighbors(nodes, fanouts, NonUniformPickUDF);
```

And the each UDF will handle the details such as whether it falls back to picking all neighbors, etc. For example,

```c++
void UniformPickUDF(...) {
  // You can bind any variables to it that the `SampleNeighbors` function does not need to know.

  if ((fanout == -1) || (num_neighbors <= fanout && !replace) {
    do_the_fast_path();
  } else {
    do_the_normal_work();
  }
}

void NonUniformPickUDF(...) {
  // You can bind any variables to it that the `SampleNeighbors` function does not need to know.
  auto positive_prop = (probs.value().slice(0, offset, offset + num_neighbors) > 0);
  if ((fanout == -1) || (num_neighbors <= positive_prop.size() && !replace) {
    // Note you could have different fast path condition than uniform sampling.
    do_the_fast_path();
  } else {
    do_the_normal_work();
  }
}
```

Reply:
The central idea is that if we believe the function `SampleNeighbors` only handles the parallelism of outer-loop as well as locating the segment belonging to the same type, then arguments like `replace`, `prob`, `fanouts` are actually all irrelevant.

Reply:
If you think this requires too much a change, one way to quickly makes it read better is to restructure this `Pick` function so that uniform and non-uniform sampling logics are in two blocks. E.g.,

```c++
torch::Tensor Pick(
    int64_t offset, int64_t num_neighbors, int64_t fanout, bool replace,
    const torch::TensorOptions& options,
    const torch::optional<torch::Tensor>& probs) {
  if (probs.has_value()) {
    return NonUniformPick(offset, num_neighbors, fanout, replace, options, probs);
  } else {
    return UniformPick(offset, num_neighbors, fanout, replace, options, probs);
  }
}
```

Reply:
Make sense, the design of using a UDF make the code and logic more clear, as well as help clarify the responsibility of the 'sample_neighbors' code block, It is a good choice. But as the change is too much, and as the principle of 'PR should be short and only do one thing', we can first adopt the 'quick way' to add the probs logic and make the refactor in the future.